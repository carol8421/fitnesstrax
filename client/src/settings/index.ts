export * from "./languages"
export * from "./unitSystems"
