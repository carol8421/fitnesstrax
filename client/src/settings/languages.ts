import _ from "lodash/fp"
import { first } from "../common"

import Option from "../option"

export class Language {
  constructor(readonly sym: string, readonly name: string) {}
}

export const English = new Language("en", "English")
export const Esperanto = new Language("eo", "Esperanto")

export const Languages: Array<Language> = [English, Esperanto]

export const languageFromSymbol = (s: string): Option<Language> =>
  first(_.filter((lang: Language): boolean => lang.sym === s)(Languages))
