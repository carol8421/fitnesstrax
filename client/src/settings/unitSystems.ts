import { Option } from "ld-ambiguity"
import _ from "lodash/fp"

import { first } from "../common"
import * as msgs from "../translations"

export class UnitSystem {
  constructor(
    readonly sym: string,
    readonly length: string,
    readonly lengthRepr: msgs.Message,
    readonly mass: string,
    readonly massRepr: msgs.Message,
  ) {}
}

export const SI = new UnitSystem(
  "SI",
  "km",
  msgs.kilometers,
  "kg",
  msgs.kilograms,
)
export const USA = new UnitSystem(
  "USA",
  "miles",
  msgs.miles,
  "lbs",
  msgs.pounds,
)
export const UK = new UnitSystem(
  "UK",
  "miles",
  msgs.miles,
  "stones",
  msgs.stones,
)
export const Canada = new UnitSystem(
  "Canada",
  "km",
  msgs.kilometers,
  "kg",
  msgs.kilograms,
)

export const UnitSystems: Array<UnitSystem> = [SI, USA, UK, Canada]

export const unitsystemFromSymbol = (s: string): Option<UnitSystem> =>
  first(_.filter((lang: UnitSystem): boolean => lang.sym === s)(UnitSystems))
