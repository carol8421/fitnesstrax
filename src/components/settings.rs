use chrono_tz::Tz;
use gtk::prelude::*;
use std::cell::RefCell;
use std::path::PathBuf;
use std::rc::Rc;
use std::sync::{Arc, RwLock};

use crate::components::{
    dropmenu_c, labeled_widget_c, text_entry_c, Component, Container, MenuOptions,
};
use crate::context::Application;
use crate::i18n::{Text, UnitSystem};

#[derive(Clone)]
pub struct Settings {
    widget: gtk::Box,

    database_path_widget: Container,
    language_widget: Container,
    timezone_widget: Container,
    units_widget: Container,

    ctx: Arc<RwLock<Application>>,
}

impl Settings {
    pub fn new(ctx: Arc<RwLock<Application>>) -> Settings {
        let widget = gtk::Box::new(gtk::Orientation::Vertical, 5);

        let no_widget: Option<gtk::Widget> = None;

        let mut component = Settings {
            widget: widget.clone(),
            database_path_widget: Container::new(no_widget.clone()),
            language_widget: Container::new(no_widget.clone()),
            timezone_widget: Container::new(no_widget.clone()),
            units_widget: Container::new(no_widget.clone()),
            ctx: ctx.clone(),
        };

        component
            .widget
            .pack_start(&component.database_path_widget.widget, false, false, 0);
        component
            .widget
            .pack_start(&component.language_widget.widget, false, false, 0);
        component
            .widget
            .pack_start(&component.timezone_widget.widget, false, false, 0);
        component
            .widget
            .pack_start(&component.units_widget.widget, false, false, 0);

        let (series_path, settings) = {
            let ctx = ctx.read().unwrap();
            let state = ctx.get_state();

            (
                state.series_path().map(|v| v.clone()),
                state.settings().clone(),
            )
        };

        {
            let ctx = ctx.clone();
            let chooser =
                gtk::FileChooserButton::new("database file", gtk::FileChooserAction::Open);
            if let Some(sp) = series_path {
                chooser.set_filename(sp);
            }
            chooser.connect_file_set(move |chooser| {
                if let Some(filename) = chooser.get_filename() {
                    ctx.write().unwrap().set_series_path(filename);
                }
            });
            component.database_path_widget.swap(Some(labeled_widget_c(
                &settings.text.database_path(),
                chooser,
            )));
        }

        {
            let component = Rc::new(RefCell::new(component.clone()));

            component
                .borrow_mut()
                .language_widget
                .swap(Some(language_menu(&settings.text, component.clone())));
            component
                .borrow_mut()
                .timezone_widget
                .swap(Some(timezone_menu(
                    &settings.text,
                    &settings.timezone,
                    component.clone(),
                )));
            component.borrow_mut().units_widget.swap(Some(units_menu(
                &settings.text,
                &settings.units,
                component.clone(),
            )));
        }

        component.widget.show_all();

        component
    }

    fn set_language(&mut self, language: &str) {
        let (series_path, text, timezone, units) = {
            let mut ctx = self.ctx.write().unwrap();
            ctx.set_language(language);
            let state = ctx.get_state();
            let settings = state.settings();
            (
                state
                    .series_path()
                    .and_then(|s| s.to_str())
                    .map(String::from),
                settings.text.clone(),
                settings.timezone,
                settings.units,
            )
        };

        let ctx = self.ctx.clone();
        self.database_path_widget.swap(Some(labeled_widget_c(
            &text.database_path(),
            text_entry_c(
                &series_path.unwrap_or(String::from("")),
                Box::new(move |s| ctx.write().unwrap().set_series_path(PathBuf::from(s))),
            ),
        )));

        {
            let component = Rc::new(RefCell::new(self.clone()));
            self.language_widget
                .swap(Some(language_menu(&text, component.clone())));
            self.timezone_widget
                .swap(Some(timezone_menu(&text, &timezone, component.clone())));
            self.units_widget
                .swap(Some(units_menu(&text, &units, component.clone())));
        }
    }

    fn set_timezone(&self, timezone_str: &str) {
        let mut ctx = self.ctx.write().unwrap();
        ctx.set_timezone(timezone_str.parse().unwrap());
    }

    fn set_units(&self, units: &str) {
        let mut ctx = self.ctx.write().unwrap();
        ctx.set_units(units);
    }
}

impl Component for Settings {
    fn widget(&self) -> gtk::Widget {
        self.widget.clone().upcast::<gtk::Widget>()
    }
}

fn language_menu(text: &Text, component: Rc<RefCell<Settings>>) -> gtk::Widget {
    labeled_widget_c(
        text.language().as_str(),
        dropmenu_c(
            MenuOptions(vec![("en", "English"), ("eo", "Esperanto")]),
            text.language_id().get_language(),
            Box::new(move |s| component.borrow_mut().set_language(s)),
        ),
    )
}

fn timezone_menu(text: &Text, timezone: &Tz, component: Rc<RefCell<Settings>>) -> gtk::Widget {
    labeled_widget_c(
        &text.timezone(),
        dropmenu_c(
            MenuOptions(vec![
                ("America/Chicago", "United States: Chicago"),
                ("America/New_York", "United States: New York"),
            ]),
            timezone.name(),
            Box::new(enclose!(component => move |s| component.borrow_mut().set_timezone(s))),
        ),
    )
}

fn units_menu(text: &Text, units: &UnitSystem, component: Rc<RefCell<Settings>>) -> gtk::Widget {
    labeled_widget_c(
        &text.units(),
        dropmenu_c(
            MenuOptions(vec![
                ("SI", "SI (kg, km, m/s)"),
                ("USA", "USA (lbs, mi, mph)"),
            ]),
            &String::from(units),
            Box::new(enclose!(component => move |s| component.borrow_mut().set_units(s))),
        ),
    )
}
