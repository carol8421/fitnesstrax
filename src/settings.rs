use chrono_tz::Tz;
use std::convert::TryFrom;
use unic_langid::LanguageIdentifier;

use crate::config::Configuration;
use crate::i18n::{Text, UnitSystem};

#[derive(Clone, Debug)]
pub struct Settings {
    pub timezone: Tz,
    pub units: UnitSystem,
    pub text: Text,
}

impl Settings {
    pub fn new(langid: LanguageIdentifier, units: UnitSystem, timezone: chrono_tz::Tz) -> Settings {
        let text = Text::new(langid);

        Settings {
            timezone,
            units,
            text,
        }
    }

    pub fn from_config(config: &Configuration) -> Settings {
        Settings::new(
            (&config.language).into(),
            config.units.clone(),
            config.timezone,
        )
    }

    pub fn set_language(&mut self, lang_str: &str) {
        let langid = lang_str.parse().expect("Language parsing failed");
        self.text = Text::new(langid);
    }

    pub fn set_units(&mut self, units_str: &str) {
        self.units = UnitSystem::try_from(units_str).expect("invalid unit system identifier");
    }

    pub fn set_timezone(&mut self, timezone: chrono_tz::Tz) {
        self.timezone = timezone;
    }
}
