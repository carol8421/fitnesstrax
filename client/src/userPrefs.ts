import { IANAZone } from "luxon"

import Option from "./option"
import { Language, UnitSystem } from "./settings"

export class UserPreferences {
  constructor(
    readonly timezone: IANAZone,
    readonly units: UnitSystem,
    readonly language: Language,
  ) {}

  withTimezone = (tz: IANAZone): UserPreferences =>
    new UserPreferences(tz, this.units, this.language)

  withUnits = (units: UnitSystem): UserPreferences =>
    new UserPreferences(this.timezone, units, this.language)

  withLanguage = (lang: Language): UserPreferences =>
    new UserPreferences(this.timezone, this.units, lang)
}
