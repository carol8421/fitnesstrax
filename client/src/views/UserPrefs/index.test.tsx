import React from "react"
import { shallow } from "enzyme"
import { IANAZone } from "luxon"

import { English, SI } from "../../settings"
import { UserPreferences } from "../../userPrefs"
import UserPreferenceView from "./index"

describe("UserPreferencesView", () => {
  it("renders with default values", () => {
    const prefs = new UserPreferences(IANAZone.create("UTC"), SI, English)
    const wrapper = shallow(
      <UserPreferenceView prefs={prefs} onSave={() => null} />,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
