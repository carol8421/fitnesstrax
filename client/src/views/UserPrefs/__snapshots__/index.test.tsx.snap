// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPreferencesView renders with default values 1`] = `ShallowWrapper {}`;
